### PR TITLE
Refactor getRule to staticText variable for Abilities

### PR DESCRIPTION
**Pull request by Alex-Vasile**
One less method to implement for custom abilities so less boilerplate code needed for new cards.

This PR has the initial implementation, all of the custom abilities for cards starting with A.

### CODE REVIEW

**Comment by JayDi85** [comment]
GetRule is dynamic method, there are possible changes for card texts in the
game (not only in construction, also multiple constructions possible).

Must make more research before starting to change it (must take GetRule and
try to find any non text only code).

**Comment by JayDi85** [comment]
ok, I've got your idea -- remove method usage from some cards/abilities. Not `GetRule` method at all. Then it's fine.

**Comment by Alex-Vasile** [comment]
> ok, I've got your idea -- remove method usage from some cards/abilities. Not `GetRule` method at all. Then it's fine.
> 
> GetRule must exists as dynamic method: shot_220926_071428

Yes. Sorry for not being more clear.

My plan is to remove Overrides to get rule whenever possible, and instead store the value in staticText, the same way that it's done for effects.

File: Mage.Sets/src/mage/cards/a/AmuletOfVigor.java [hunk 1]
```diff
@@ -76,6 +76,7 @@ public boolean checkTrigger(GameEvent event, Game game) {
     @Override
     public String getRule() {
         // that triggers depends on stack order, so make each trigger unique with extra info
+        // TODO: Do the hints belong here? Must be a better place for them.
```
Comment (Alex-Vasile) [inline Mage.Sets/src/mage/cards/a/AmuletOfVigor.java:79]: Maybe take similar approach to https://github.com/magefree/mage/commit/07902f64b77da784fb23b136324314e7540c7f99?

**Comment by Alex-Vasile** [comment]
Failing test is:
```
[ERROR] org.mage.test.lki.LastKnownInformationTest.testPersistTriggersInTime  Time elapsed: 35.521 s  <<< FAILURE!
java.lang.AssertionError: 
Found errors on card loading: 
Error loading card: mage.cards.h.HallowedHealer (ODY)
Error loading card: mage.cards.k.KitsuneBonesetter (SOK)
	at org.mage.test.lki.LastKnownInformationTest.<init>(LastKnownInformationTest.java:12)
```

Both of those use `ConditionalActivatedAbility` which was one of the common abilities modified here. I am not sure why they would be failing the LKI test... will investigate

**Comment by awjackson** [comment]
With this change, you are now calling super.getRule() statically in the constructor of `ConditionalActivatedAbility` rather than dynamically. This will fail if any `Effect`s of the `ConditionalActivatedAbility` have dynamic text (generally targetted effects do),

**Comment by Alex-Vasile** [comment]
> With this change, you are now calling super.getRule() statically in the constructor of `ConditionalActivatedAbility` rather than dynamically. This will fail if any `Effect`s of the `ConditionalActivatedAbility` have dynamic text (generally targetted effects do), It will also fail if effects are added to the ability after it is initially created (i.e. any time the ability needs more than one effect). You can't statically generate the rules text of any ability that can have arbitrary effects added to it.

Yup. I think that should be fixed. How's that.

**Comment by Alex-Vasile** [comment]
I have removed all of the changes that were not part of custom abilities (or the base Ability class)

**Comment by Alex-Vasile** [comment]
The latest failure is caused by the Verify test thinking that Ashenmoor Liege has ward.

This is a strange given that I haven't change the content of the text, only that it's now stored in a string which is returned by the super.getRule()....

**Comment by awjackson** [comment]
Ashenmoor Liege's ability extends `TriggeredAbilityImpl`, not `AbilityImpl`. The `getRule` of `TriggeredAbilityImpl` is what is being called, and that does not check for static text.

**Comment by Alex-Vasile** [comment]
After the discussion in #8973 about improvements to the game engine for better performance, I am convinced that this PR would be a step in the wrong direction.

This would add yet another piece of state to the abilities, which would be counter productive to the approach I outlined in that issue.

**Pull request closed without merging**